File: cs50/week0to5-c/pset2/readability/readability.c
#include <stdio.h>
#include <cs50.h>
#include <string.h>
#include <ctype.h>
#include <math.h>

string format_lower(string text);
int find_words(string text);
int find_sentences(string text);
int find_letters(string text);


int main(void)
{
    // Get input text from user
    string user_text;
    do
    {
        user_text = get_string("Text: ");
    }
    while (strlen(user_text) == 0);

    string formatted_text = format_lower(user_text);
    //printf("lower text: %s\n", formatted_text);


    // Find the number of words, sentences and letters in the user text
    int words = find_words(formatted_text);
    //printf("Word count: %i\n", words);

    int letters = find_letters(formatted_text);
    //printf("Letter count: %i\n", letters);

    int sentences = find_sentences(formatted_text);
    //printf("Sentence count: %i\n", sentences);

    // Calculating necessary l and s values
    float l = (float) letters / words * 100.00; // calculate letters per 100 words
    //printf("Letters per 100 words: %f\n", l);
    float s = (float) sentences / words * 100.00; // calculate sentences per 100 words
    //printf("Sentences per 100 words: %f\n", s);

    // Use the coleman index formula
    int grade = round(0.0588 * l - 0.296 * s - 15.8);


    // Output the index to the user
    if (grade < 16 && grade > 1)
    {
        printf("Grade %i\n", grade);
    }
    else if (grade >= 16)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Before Grade 1\n");
    }
}


string format_lower(string text)
{

    for (int i = 0, n = strlen(text); i < n; i++)
    {
        text[i] = tolower(text[i]);
    }
    return text;
}


int find_words(string text)
{
    // Find the number of words in the text
    int counter = 1;
    for (int i = 0, n = strlen(text); i < n; i++)
    {
        if (text[i] == ' ')
        {
            counter++;
        }
    }

    return counter;

}

int find_sentences(string text)
{
    // Checking for valid string
    int length = strlen(text);
    if (length == 0)
    {
        return 0;
    }

    // Find the number of words in the text

    int counter = 0;
    for (int i = 0; i < length; i++)
    {
        if (text[i] == '.' || text[i] == '!' || text[i] == '?')
        {
            counter++;
        }
    }

    if (counter == 0) // in case there are no punctuations, setting the value of counter to 1
    {
        counter = 1;
    }
    return counter;
}

int find_letters(string text)
{
    // Checking for valid string
    int length = strlen(text);
    if (length == 0)
    {
        return 0;
    }


    // TODO: Find the number of letters in the text
    int counter = 0;
    for (int i = 0; i < length; i++)
    {
        if (isalpha(text[i]))
        {
            counter++;
        }
    }
    return counter;
}

File: cs50/week6to10-python/pset6/readability/readability.py
import re


def main():
    # get input text from user
    user_text = input("Text: ")

    # call functions to get number of sentences, words and letters in the text
    s = sentences(user_text)
    l = letters(user_text)
    words = word_count(user_text)

    # sentences and letters per 100 words
    s = s / words * 100
    print("s: " + str(s))
    l = l / words * 100
    print("l: " + str(l))

    # calculate coleman lieu index
    index = 0.0588 * l - 0.296 * s - 15.8

    # print the necessary indexstyle50 readability.py
    if index > 1 and index < 16:
        print(f"Grade {round(index)}")
    elif index < 1:
        print("Before Grade 1")
    else:
        print("Grade 16+")


def sentences(text):
    '''
    returns the number of sentences per 100 words
    input: text
    output: sentences in the text
    '''
    return len(re.split(r'[.!?]+', text)) - 1


def letters(text):
    '''
    returns the number of letters per 100 words
    input: text
    output: number of letters in text without puncutations and spaces
    '''
    letter_count = 0
    text = re.sub(r'[^\w\s]', '', text)
    text = text.replace(" ", "")

    return len(text)


def word_count(text):
    '''
    returns the word count of a text
    input: user text
    output: count of words in text

    '''
    return len(text.split())
    

if __name__ == "__main__":
    main()